components/proxy/src/test/java/com/hotels/styx/admin/handlers/ServiceProviderHandlerTest.java

```diff
@@ -0,0 +1,168 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.admin.handlers;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.hotels.styx.api.HttpRequest;
+import com.hotels.styx.api.HttpResponse;
+import com.hotels.styx.api.HttpResponseStatus;
+import com.hotels.styx.api.extension.service.spi.StyxService;
+import com.hotels.styx.routing.config.StyxObjectDefinition;
+import com.hotels.styx.routing.db.StyxObjectStore;
+import com.hotels.styx.routing.handlers.ProviderObjectRecord;
+import com.hotels.styx.server.HttpInterceptorContext;
+import org.apache.commons.lang.StringUtils;
+import org.mockito.Mockito;
+import org.testng.annotations.Test;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.testng.Assert.assertTrue;
+
+public class ServiceProviderHandlerTest {
+
+    @Test
+    public void returnsAllProviders() throws IOException {
+        StyxObjectStore<ProviderObjectRecord> store = createTestStore();
+        ServiceProviderHandler handler = new ServiceProviderHandler(store);
+        HttpRequest request = HttpRequest.get("/admin/service/providers").build();
+        HttpResponse response = Mono.from(handler.handle(request, HttpInterceptorContext.create())).block();
+
+        assertThat(response.status(), equalTo(HttpResponseStatus.OK));
+
+        List<StyxObjectDefinition> actualProviders = extractProviders(response.bodyAs(StandardCharsets.UTF_8));
+        assertThat(actualProviders.size(), equalTo(store.entrySet().size()));
+        for (StyxObjectDefinition actual :
+                actualProviders) {
+            Optional<ProviderObjectRecord> rec = store.get(actual.name());
+            assertTrue(rec.isPresent());
+            validateProvider(actual, rec.get());
+        }
+    }
+
+    @Test
+    public void returnsNoContentStatusWhenNoProvidersAvailable() {
+        StyxObjectStore<ProviderObjectRecord> empty = new StyxObjectStore<>();
+        ServiceProviderHandler handler = new ServiceProviderHandler(empty);
+        HttpRequest request = HttpRequest.get("/admin/service/providers").build();
+        HttpResponse response = Mono.from(handler.handle(request, HttpInterceptorContext.create())).block();
+
+        assertThat(response.status(), equalTo(HttpResponseStatus.NO_CONTENT));
+        assertThat(response.contentLength().get(), equalTo(0L));
```

Comment:
I don't particularly like the Exception message when we call get() on an optional. Maybe we could have a nicer assertion message.

Reply:
Perhaps assert `isPresent()` beforehand?

Reply:
Ah, actually it seems a NO_CONTENT response should not have a content-length header at all, even one set to 0. So I'll fix that, and check for its absence.
https://tools.ietf.org/html/rfc7230#section-3.3.2

> A server MUST NOT send a Content-Length header field in any response
   with a status code of 1xx (Informational) or 204 (No Content).

Reply:
I think we intentionally let it have  "Content-Length: 0" here as it seems otherwise the client was getting  'Transfer-Encoding: chunked' otherwise and that broke everything apart.  If I recall correctly, we thought the header was added by Netty but we didn't really verify it.

Reply:
In a quick check of the running server, I didn't get a `Transfer-Encoding: chunked` header when a zero-content response was returned. Maybe the admin server's netty is set up differently to the main proxy server?